File: packages/golfie_core/metrics/units.py
"""Unit conversions between Golfie's internal metric units and the
yards/mph/degrees/rpm units the UI displays (spec section 6).

Internal storage is always meters/seconds/radians; convert only at the
edges (API serialization for display, or user-facing exports).
"""

from __future__ import annotations

import math

from golfie_core.config.physical_constants import METERS_PER_YARD, MPS_PER_MPH


def mps_to_mph(value_mps: float) -> float:
    return value_mps / MPS_PER_MPH


def mph_to_mps(value_mph: float) -> float:
    return value_mph * MPS_PER_MPH


def meters_to_yards(value_m: float) -> float:
    return value_m / METERS_PER_YARD


def yards_to_meters(value_yd: float) -> float:
    return value_yd * METERS_PER_YARD


def rad_to_deg(value_rad: float) -> float:
    return math.degrees(value_rad)


def deg_to_rad(value_deg: float) -> float:
    return math.radians(value_deg)

File: apps/web/frontend/src/lib/units.ts
// Mirrors packages/golfie_core/metrics/units.py. Internal/API values are
// always meters/seconds/m-per-second; the UI converts to yards/mph at
// the display edge only (spec section 6).

const METERS_PER_YARD = 0.9144;
const MPS_PER_MPH = 0.44704;

export const mpsToMph = (mps: number): number => mps / MPS_PER_MPH;
export const metersToYards = (m: number): number => m / METERS_PER_YARD;

export function formatMetric(value: number | null, unit: "mph" | "yd" | "deg" | "rpm" | "x"): string {
  if (value === null) return "—";
  switch (unit) {
    case "mph":
      return `${mpsToMph(value).toFixed(1)} mph`;
    case "yd":
      return `${metersToYards(value).toFixed(1)} yd`;
    case "deg":
      return `${value.toFixed(1)}°`;
    case "rpm":
      return `${value.toFixed(0)} rpm`;
    case "x":
      return value.toFixed(2);
    default:
      return `${value}`;
  }
}
